src/jsn.erl

```diff
@@ -357,6 +359,71 @@ path_elements(_Type, _Path, _Acc) -> erlang:error(badarg).
 
 
 %%==============================================================================
+%% select API
+%%==============================================================================
+
+%%------------------------------------------------------------------------------
+%% @doc Transforms a list of Elements according to a given OutputSpecification
+%%------------------------------------------------------------------------------
+-spec select(selection(), json_array()) -> json_array().
+select(OutputSpec, Elements) when is_list(Elements) ->
+    select(OutputSpec, [], Elements);
+select(_OutpuSpec, _Elements) ->
+    erlang:error(badarg).
+
+%%------------------------------------------------------------------------------
+%% @doc Same as select/2 but it first filters the elements according to a given
+%% list of Conditions
+%%------------------------------------------------------------------------------
+-spec select(selection(), conditions(), json_array()) -> json_array().
+select(OutputSpec, Conditions, Elements) when is_list(Elements) ->
+    %% ensures Conditions are a list
+    ConditionList =
+        case Conditions of
+            _ when is_list(Conditions) -> Conditions;
+            _ -> [Conditions]
+        end,
+    %% This function applies a single condition to a single element
+    ApplyConditionFun = 
+        fun({Path, FilterFun}, E) when is_function(FilterFun) ->
+                erlang:apply(FilterFun, [?MODULE:get(Path, E)]);
+           ({Path, Value}, E) ->
+                ?MODULE:get(Path, E) == Value;
+           (FilterFun, E) when is_function(FilterFun) ->
+                erlang:apply(FilterFun, [E]);
+           (_, _) ->
+                erlang:error(badarg)
+        end,
+    FilteredElements =
+        lists:foldl(fun(Condition, CurrentElements) ->
+                        lists:filter(fun(E) ->
+                                        erlang:apply(ApplyConditionFun, [Condition, E])
+                                    end, CurrentElements)
+                    end, Elements, ConditionList),
+    %% Converts Filtered Elements by applying output specifications
+    OutputFun =
+        case OutputSpec of
+            identity ->
+                fun(E) -> E end;
+            {value, Path} ->
+                fun(E) -> ?MODULE:get(Path, E) end;
+            {value, Path, Default} ->
+                fun(E) -> ?MODULE:get(Path, E, Default) end;
+            _ when is_list(OutputSpec) ->
+                fun(E) ->
+                    lists:map(fun({value, Path}) -> ?MODULE:get(Path, E);
+                                 ({value, Path, Default}) -> ?MODULE:get(Path, E, Default)
+                              end, OutputSpec)
+                end;
+            _ ->
+                erlang:error(badarg)
```

Comment:
Same issue here W.R.T. traceable exceptions. Either this should include the bad value that triggered the input, or it should fail at a function level where the immediate argument that triggered the failure is clear.

Reply:
Really agree with this sentiment. Make it easy for users to understand specifically what's bad/invalid when a function dies.